# In-Class Coding Lab: Data Analysis with Pandas

In this lab, we will perform a data analysis on the **RMS Titanic** passenger list. The RMS Titanic is one of the most famous ocean liners in history. On April 15, 1912 it sank after colliding with an iceberg in the North Atlantic Ocean. To learn more, read here: https://en.wikipedia.org/wiki/RMS_Titanic 

Our goal today is to perform a data analysis on a subset of the passenger list. We're looking for insights as to which types of passengers did and didn't survive. Women? Children? 1st Class Passengers? 3rd class? Etc. 

I'm sure you've heard the expression often said during emergencies: "Women and Children first" Let's explore this data set and find out if that's true!

Before we begin you should read up on what each of the columns mean in the data dictionary. You can find this information on this page: https://www.kaggle.com/c/titanic/data 


## Loading the data set

First we load the dataset into a Pandas `DataFrame` variable. The `sample(10)` method takes a random sample of 10 passengers from the data set.

In [46]:
import pandas as pd
import numpy as np

# this turns off warning messages
import warnings
warnings.filterwarnings('ignore')

passengers = pd.read_csv('CCL-titanic.csv')
passengers.sample(10)

PassengerId  Survived  Pclass  \
442          443         0       3   
427          428         1       2   
44            45         1       3   
344          345         0       2   
69            70         0       3   
59            60         0       3   
629          630         0       3   
490          491         0       3   
311          312         1       1   
143          144         0       3   

                                                  Name     Sex   Age  SibSp  \
442                          Petterson, Mr. Johan Emil    male  25.0      1   
427  Phillips, Miss. Kate Florence ("Mrs Kate Louis...  female  19.0      0   
44                       Devaney, Miss. Margaret Delia  female  19.0      0   
344                            Fox, Mr. Stanley Hubert    male  36.0      0   
69                                   Kink, Mr. Vincenz    male  26.0      2   
59                  Goodwin, Master. William Frederick    male  11.0      5   
629                           O'Connell, Mr. Patrick D    male   NaN      0   
490               Hagland, Mr. Konrad Mathias Reiersen    male   NaN      1   
311                         Ryerson, Miss. Emily Borie  female  18.0      2   
143                                Burke, Mr. Jeremiah    male  19.0      0   

     Parch    Ticket      Fare            Cabin Embarked  
442      0    347076    7.7750              NaN        S  
427      0    250655   26.0000              NaN        S  
44       0    330958    7.8792              NaN        Q  
344      0    229236   13.0000              NaN        S  
69       0    315151    8.6625              NaN        S  
59       2   CA 2144   46.9000              NaN        S  
629      0    334912    7.7333              NaN        Q  
490      0     65304   19.9667              NaN        S  
311      2  PC 17608  262.3750  B57 B59 B63 B66        C  
143      0    365222    6.7500              NaN        Q

## How many survived?

One of the first things we should do is figure out how many of the passengers in this data set survived. Let's start with isolating just the `'Survivied'` column into a series:

In [4]:
passengers['Survived'].sample(10)

301    1
231    0
475    0
361    0
48     0
34     0
430    1
389    1
582    0
479    1
Name: Survived, dtype: int64

There's too many to display so we just display a random sample of 10 passengers. 

- 1 means the passenger survivied
- 0 means the passenger died

What we really want is to count the number of survivors and deaths. We do this by querying the `value_counts()` of the `['Survived']` column, which returns a `Series` of counts, like this:

In [47]:
passengers['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

Only 342 passengers survived, and 549 perished. Let's observe this same data as percentages of the whole. We do this by adding the `normalize=True` named argument to the `value_counts()` method.

In [49]:
passengers['Survived'].value_counts(normalize=True)*100

0    61.616162
1    38.383838
Name: Survived, dtype: float64

**Just 38% of passengers in this dataset survived.**

### Now you Try it!

**FIRST** Write a Pandas expression to display counts of males and female passengers using the `Sex` variable:

In [51]:
# todo write code here
passengers['Sex'].value_counts()
#urvivePercentage= surviveIndex*100
#print("Just %.0f of paasengers ")

male      577
female    314
Name: Sex, dtype: int64

**NEXT** Write a Pandas expression to display male /female passenger counts as a percentage of the whole number of passengers in the data set.

In [52]:
# todo write code here
passengers['Sex'].value_counts(normalize=True)*100

male      64.758698
female    35.241302
Name: Sex, dtype: float64

If you got things working, you now know that **35% of passengers were female**.

## Who survivies? Men or Women?

We now know that 35% of the passengers were female, and 65% we male. 

**The next think to think about is how do survivial rates affect these numbers? **

If the ratio is about the same for surviviors only, then we can conclude that your **Sex** did not play a role in your survival on the RMS Titanic. 

Let's find out.

In [53]:
survivors = passengers[passengers['Survived'] ==1]
survivors['PassengerId'].count()

342

Still **342** like we discovered originally. Now let's check the **Sex** split among survivors only:

In [54]:
survivors['Sex'].value_counts()

female    233
male      109
Name: Sex, dtype: int64

WOW! That is a huge difference! But you probably can't see it easily. Let's represent it in a `DataFrame`, so that it's easier to visualize:

In [55]:
sex_all_series = passengers['Sex'].value_counts()
sex_survivor_series = survivors['Sex'].value_counts()

sex_comparision_df = pd.DataFrame({ 'AllPassengers' : sex_all_series, 'Survivors' : sex_survivor_series })
sex_comparision_df['SexSurvivialRate'] = sex_comparision_df['Survivors'] / sex_comparision_df['AllPassengers']
sex_comparision_df

AllPassengers  Survivors  SexSurvivialRate
female            314        233          0.742038
male              577        109          0.188908

 **So, females had a 74% survival rate. Much better than the overall rate of 38%**
 
We should probably briefly explain the code above. 

- The first two lines get a series count of all passengers by Sex (male / female) and count of survivors by sex
- The third line creates DataFrame. Recall a pandas dataframe is just a dict of series. We have two keys 'AllPassengers' and 'Survivors'
- The  fourth line creates a new column in the dataframe which is just the survivors / all passengers to get the rate of survival for that Sex.

## Feature Engineering: Adults and Children

Sometimes the variable we want to analyze is not readily available, but can be created from existing data. This is commonly referred to as **feature engineering**. The name comes from machine learning where we use data called *features* to predict an outcome. 

Let's create a new feature called `'AgeCat'` as follows:

- When **Age** <=18 then 'Child'
- When **Age** >18 then 'Adult'

This is easy to do in pandas. First we create the column and set all values to `np.nan` which means 'Not a number'. This is Pandas way of saying no value. Then we set the values based on the rules we set for the feature.

In [56]:
passengers['AgeCat'] = np.nan # Not a number
passengers['AgeCat'][ passengers['Age'] <=18 ] = 'Child'
passengers['AgeCat'][ passengers['Age'] > 18 ] = 'Adult'
passengers.sample(5)

PassengerId  Survived  Pclass                                      Name  \
462          463         0       1                         Gee, Mr. Arthur H   
630          631         1       1      Barkworth, Mr. Algernon Henry Wilson   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
822          823         0       1           Reuchlin, Jonkheer. John George   
634          635         0       3                        Skoog, Miss. Mabel   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked AgeCat  
462    male  47.0      0      0      111320  38.50   E63        S  Adult  
630    male  80.0      0      0       27042  30.00   A23        S  Adult  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S    NaN  
822    male  38.0      0      0       19972   0.00   NaN        S  Adult  
634  female   9.0      3      2      347088  27.90   NaN        S  Child

Let's get the count and distrubutions of Adults and Children on the passenger list.

In [57]:
passengers['AgeCat'].value_counts()

Adult    575
Child    139
Name: AgeCat, dtype: int64

And here's the percentage as a whole:

In [59]:
passengers['AgeCat'].value_counts(normalize=True)*100

Adult    80.532213
Child    19.467787
Name: AgeCat, dtype: float64

So close to **80%** of the passengers were adults. Once again let's look at the ratio of `AgeCat` for survivors only. If your age has no bearing of survivial, then the rates should be the same. 

Here's the counts of Adult / Children among the survivors only:

In [67]:
survivors = passengers[passengers['Survived'] ==1]
survivors['AgeCat'].value_counts()

Adult    220
Child     70
Name: AgeCat, dtype: int64

### Now You Try it!

Calculate the `AgeCat` survival rate, similar to how we did for the `SexSurvivalRate`. 

In [68]:
agecat_all_series = passengers['AgeCat'].value_counts()
agecat_survivor_series = survivors['AgeCat'].value_counts()


# todo make a data frame, add AgeCatSurvivialRate column, display dataframe 

#passengers['AgeCatSurvivialRate'] = 'No'
#passengers['AgeCatSurvivialRate'] = 'Yes'
#passengers


agecat_comparision_df = pd.DataFrame({ 'AllPassengers' : agecat_all_series, 'Survivors' : agecat_survivor_series })
agecat_comparision_df['AgeCatSurvivialRate'] = agecat_comparision_df['Survivors']*100 / agecat_comparision_df['AllPassengers']
agecat_comparision_df


AllPassengers  Survivors  AgeCatSurvivialRate
Adult            575        220            38.260870
Child            139         70            50.359712

**So, children had a 50% survival rate, better than the overall rate of 38%**

## So, women and children first?

It looks like the RMS really did have the motto: "Women and Children First."

Here's our insights. We know:

- If you were a passenger, you had a 38% chance of survival.
- If you were a female passenger, you had a 74% chance of survival.
- If you were a child passenger, you had a 50% chance of survival. 


### Now you try it for Passenger Class

Repeat this process for `Pclass` The passenger class variable. Display the survival rates for each passenger class. What does the information tell you about passenger class and survival rates?

I'll give you a hint... "Money Talks"


In [70]:
# todo: repeat the analysis in the previous cell for Pclass 


Pclass_all_series = passengers['Pclass'].value_counts()
Pclass_survivor_series = survivors['Pclass'].value_counts()


Pclass_comparision_df = pd.DataFrame({ 'AllPassengers' : Pclass_all_series, 'Survivors' : Pclass_survivor_series })
Pclass_comparision_df['PclassSurvivialRate'] = Pclass_comparision_df['Survivors']*100 / Pclass_comparision_df['AllPassengers']
Pclass_comparision_df

AllPassengers  Survivors  PclassSurvivialRate
1            216        136            62.962963
2            184         87            47.282609
3            491        119            24.236253